### PR TITLE
Public API: enable other applications to start recordings.

### LINKED ISSUE #671: Provide a way to start tracking from an Automate/Tasker flow
**Is your feature request related to a problem? Please describe.**
I always forget to start tracking my sport activities, and as a result I am missing the gps tracks of glorious hikes. I would like to program my phone to automatically record my gps track when the right conditions are met.

**Describe the solution you'd like**
I would like to start recording from the App automate. This way, I can be fully in charge of determining exactly when the start/stop recording actions should be started. In my case, I would use my gps location, provided with a change of personal activity.

**Describe alternatives you've considered**
I have considered using strava, which allows to do that via the data uri `http://strava.com/nfc/record`. You can also provide the ride type in an Extras json object.

### PR DESCRIPTION
**Pull request by dennisguse**
Fixes #671.

* [x] add preference (default disabled)
* [ ] add STOP_RECORDING
* [ ] Fix TrackRecordingActivity: if it is open, it resumes the track recording after STOP_RECORDING

### CODE REVIEW

**Comment by dennisguse** [comment]
@rgmf  @pstorch Should we add a preference for this and disable it by default?
I see some potential for misuse...

**Comment by pstorch** [comment]
The user gets a notification when a recording starts, but better have an opt in preference.

**Comment by dennisguse** [comment]
@arjan-s  Moving the discussion here.

Do you know how we could react to intents that are sent while the screen is locked?
Do you have some documentation/ setup available?

https://github.com/OpenTracksApp/OpenTracks/issues/671#issuecomment-1003616024

**Comment by rgmf** [comment]
@dennisguse I don't, sorry.

What you want to do is the same that music player does, right? Could we look into the code of some player?

**Comment by arjan-s** [comment]
@dennisguse I'm afraid I don't... As I said I'm not a very experienced Android developer, so I usually search and read documentation as I go...

**Comment by dennisguse** [comment]
@arjan-s Don't worry - me too :sunglasses:

**Comment by dennisguse** [comment]
The new implementation now uses broadcast, which should be working with the screen locked.
I just have not figured out how to send an explicit broadcast (required due to API26+): https://developer.android.com/about/versions/oreo/background#broadcasts

Test application that I use is Easer and I am not sure that it supports this :/
@arjan-s Could you check it with Tasker?

**Comment by arjan-s** [comment]
That sounds very promising! I installed the debug build using Android Studio and tested it. I'm afraid I didn't get it to work either. :(
I've read the docs and sending an explicit intent should just be the case of changing the target to Broadcast and adding a Package. I believe these are the correct settings, but they don't work for me yet. :(
Action: de.dennisguse.opentracks.publicapi.START_RECORDING
Package: de.dennisguse.opentracks.debug
Target: Broadcast Receiver

**Comment by dennisguse** [comment]
So, using a BroadcastReceiver doesn't work directly as we cannot bind the service from there.
`BroadcastReceiver components are not allowed to bind to services`

**Comment by dennisguse** [comment]
Just for documentation as I will remove this commit.
Tested in API24 using `adb shell am broadcast -a START_RECORDING`.

This doesn't work:
```

    @Override
    public void onReceive(Context context, Intent intent) {
        final BroadcastReceiver.PendingResult pendingResult = goAsync();

        TrackRecordingServiceConnection.Callback serviceConnectedCallback = service -> {
            service.startNewTrack();
            pendingResult.finish();
        };

        new TrackRecordingServiceConnection(serviceConnectedCallback)
                .startAndBind(context);
    }
}
```

**Comment by dennisguse** [comment]
I checked the activity variant and it works with a locked screen (emulator and also a LineageOS device via adb).

@arjan-s Can you test this again?
I updated the README accordingly.
PS/ can only test via adb with a locked device as I have no other means to trigger something on the device.

**Comment by arjan-s** [comment]
@dennisguse Awesome, it works perfectly!
I created a Tasker task with 2 steps:

1. wait 1 minute
2. send the intent (exactly as described in the README)

After starting the task I immediately lock my phone. And perfectly after 1 minute I heard a ping that a new notification was created. The track recording works perfectly!
Thank you for your work!

**Comment by dennisguse** [comment]
@arjan-s Awesome!
Thanks for testing!
Only thing blocking it is the TrackRecordingActivity as it triggers a resume when it receives STOP_RECORDING.
If you have time to look into this that would be awesome!

**Comment by arjan-s** [comment]
Weirdly, the intent I had for starting the recording doesn't work anymore with your latest commit. The intent hasn't changed since my last test, and is still correct according to the README.
With the current version of this PR, the intent just opens the main screen of OpenTracks, but doesn't start a recording anymore.

**Comment by dennisguse** [comment]
@arjan-s Did you enable the functionality in the settings?

**Comment by arjan-s** [comment]
Doh! Now it does work! I totally missed the new setting, sorry.
Stopping the recording while in the background works fine. I'll make time to test the remaining issue tomorrow.

**Comment by arjan-s** [comment]
@dennisguse yes, I see the same behavior. When the active recording is on screen, the intent will trigger a stop and resume of the recording, instead of just stopping it.

By just reading the code, I don't see why this happens. I'd need to run with the debugger and step through the code from that point to see what really happens.
If you'd like me to do that, I'm afraid it'll have to wait until after the weekend.

I suspect it has to do with the instantiation of the Activity. Perhaps the recording is started by creating a new instance of it, which maybe also happens when STOP_RECORDING is received.

**Comment by dennisguse** [comment]
This is ready for testing.
Tested via adb commands:
```
 adb shell am start -n de.dennisguse.opentracks.debug/de.dennisguse.opentracks.publicapi.StartRecording
 adb shell am start -n de.dennisguse.opentracks.debug/de.dennisguse.opentracks.publicapi.StopRecording
```

**Comment by arjan-s** [comment]
You found the restarting problem already! That's awesome!
I can confirm it now all works exactly as advertised!

**Comment by dennisguse** [comment]
@arjan-s also your smartwatch use case?

**Comment by arjan-s** [comment]
Yes! Although I only tested with Tasker, the use case is definitely covered by this PR.
One question about that, do you think it is possible to add the option to provide a track type (hiking, cycling, etc) through an Extra field in the intent?

**Comment by dennisguse** [comment]
@arjan-s technically it's possible - might require some minor changes to internal calls.
Just go for it - open-source allows DYI :sun_with_face:

**Comment by arjan-s** [comment]
I know, that's why I love open source :)
When I start working on the Gadgetbridge code again, I'll take a shot at this.

File: src/main/res/xml/settings.xml [hunk 1]
```diff
@@ -4,56 +4,63 @@
...
         android:key="@string/settings_reset_key"
         android:negativeButtonText="@android:string/cancel"
         android:persistent="false"
         android:positiveButtonText="@android:string/ok"
-        android:title="@string/settings_all_reset"
         android:summary="@string/settings_reset_summary"
-        android:icon="@drawable/ic_baseline_autorenew_24"/>
+        android:title="@string/settings_all_reset" />
+
+    <SwitchPreferenceCompat
```
Comment (rgmf) [inline src/main/res/xml/settings.xml:53]: I'd add this inside "OpenTracks" preference. If you don't think so, then I'd add an icon next it like in the others main preferences.

**Review by rgmf — CHANGES_REQUESTED** [review]
It works for me but only if OpenTracks is opened. I don't know if it's an expected behavior ;)

File: src/main/res/xml/settings.xml (hunk 1 above)
Comment (rgmf) [reply src/main/res/xml/settings.xml:53]: BTW: fantastic feature and work ;)

Comment (arjan-s) [reply src/main/res/xml/settings.xml:53]: Weird, it works perfectly for me when OpenTracks is closed and my display is locked. Perhaps it's a battery optimization thing? I disabled that for OpenTracks.

Comment (rgmf) [reply src/main/res/xml/settings.xml:53]: > Weird, it works perfectly for me when OpenTracks is closed and my display is locked. Perhaps it's a battery optimization thing? I disabled that for OpenTracks.

Yes, I had to set battery as unrestricted for OpenTracks ;)

**Comment by pstorch** [comment]
First I tested the adb commands to start and stop a recording. Then I manually started a track recording and OT crashed:

```
2022-01-20 22:16:13.514 3042-3042/de.dennisguse.opentracks.debug D/AndroidRuntime: Shutting down VM
2022-01-20 22:16:13.515 3042-3042/de.dennisguse.opentracks.debug E/AndroidRuntime: FATAL EXCEPTION: main
    Process: de.dennisguse.opentracks.debug, PID: 3042
    java.lang.NullPointerException: Attempt to invoke interface method 'android.view.MenuItem android.view.MenuItem.setVisible(boolean)' on a null object reference
        at de.dennisguse.opentracks.TrackRecordingActivity.updateMenuItems(TrackRecordingActivity.java:329)
        at de.dennisguse.opentracks.TrackRecordingActivity.recordStopped(TrackRecordingActivity.java:368)
        at de.dennisguse.opentracks.TrackRecordingActivity.onRecordingStatusChanged(TrackRecordingActivity.java:420)
        at de.dennisguse.opentracks.TrackRecordingActivity.-$$Nest$monRecordingStatusChanged(Unknown Source:0)
        at de.dennisguse.opentracks.TrackRecordingActivity$1.lambda$onConnected$0$de-dennisguse-opentracks-TrackRecordingActivity$1(TrackRecordingActivity.java:82)
        at de.dennisguse.opentracks.TrackRecordingActivity$1$$ExternalSyntheticLambda0.onChanged(Unknown Source:4)
        at androidx.lifecycle.LiveData.considerNotify(LiveData.java:133)
        at androidx.lifecycle.LiveData.dispatchingValue(LiveData.java:146)
        at androidx.lifecycle.LiveData$ObserverWrapper.activeStateChanged(LiveData.java:468)
        at androidx.lifecycle.LiveData$LifecycleBoundObserver.onStateChanged(LiveData.java:425)
        at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.java:354)
        at androidx.lifecycle.LifecycleRegistry.addObserver(LifecycleRegistry.java:196)
        at androidx.lifecycle.LiveData.observe(LiveData.java:205)
        at de.dennisguse.opentracks.TrackRecordingActivity$1.onConnected(TrackRecordingActivity.java:82)
        at de.dennisguse.opentracks.services.TrackRecordingServiceConnection.setTrackRecordingService(TrackRecordingServiceConnection.java:144)
        at de.dennisguse.opentracks.services.TrackRecordingServiceConnection.onServiceConnected(TrackRecordingServiceConnection.java:159)
        at android.app.LoadedApk$ServiceDispatcher.doConnected(LoadedApk.java:2084)
        at android.app.LoadedApk$ServiceDispatcher$RunConnection.run(LoadedApk.java:2116)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:246)
        at android.app.ActivityThread.main(ActivityThread.java:8633)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
```

**Comment by pstorch** [comment]
Another crash after a force close and starting a recording manually in the app.

```
2022-01-20 22:22:10.423 6206-6206/de.dennisguse.opentracks.debug I/TrackRecordingService: new status RecordingStatus{trackId=null, paused=false} -> RecordingStatus{trackId=null, paused=false}
2022-01-20 22:22:10.438 6206-6206/de.dennisguse.opentracks.debug E/ActivityThread: Service de.dennisguse.opentracks.services.TrackRecordingService has leaked ServiceConnection android.speech.tts.TextToSpeech$Connection@fd07bc3 that was originally bound here
    android.app.ServiceConnectionLeaked: Service de.dennisguse.opentracks.services.TrackRecordingService has leaked ServiceConnection android.speech.tts.TextToSpeech$Connection@fd07bc3 that was originally bound here
        at android.app.LoadedApk$ServiceDispatcher.<init>(LoadedApk.java:1941)
        at android.app.LoadedApk.getServiceDispatcherCommon(LoadedApk.java:1813)
        at android.app.LoadedApk.getServiceDispatcher(LoadedApk.java:1792)
        at android.app.ContextImpl.bindServiceCommon(ContextImpl.java:1899)
        at android.app.ContextImpl.bindService(ContextImpl.java:1829)
        at android.content.ContextWrapper.bindService(ContextWrapper.java:774)
        at android.speech.tts.TextToSpeech.connectToEngine(TextToSpeech.java:849)
        at android.speech.tts.TextToSpeech.initTts(TextToSpeech.java:819)
        at android.speech.tts.TextToSpeech.<init>(TextToSpeech.java:772)
        at android.speech.tts.TextToSpeech.<init>(TextToSpeech.java:751)
        at android.speech.tts.TextToSpeech.<init>(TextToSpeech.java:735)
        at de.dennisguse.opentracks.services.announcement.VoiceAnnouncement.start(VoiceAnnouncement.java:122)
        at de.dennisguse.opentracks.services.announcement.VoiceAnnouncementManager.restore(VoiceAnnouncementManager.java:62)
        at de.dennisguse.opentracks.services.announcement.VoiceAnnouncementManager.setFrequency(VoiceAnnouncementManager.java:100)
        at de.dennisguse.opentracks.services.TrackRecordingService$2.onSharedPreferenceChanged(TrackRecordingService.java:114)
        at de.dennisguse.opentracks.settings.PreferencesUtils.registerOnSharedPreferenceChangeListener(PreferencesUtils.java:80)
        at de.dennisguse.opentracks.services.TrackRecordingService.onCreate(TrackRecordingService.java:153)
        at android.app.ActivityThread.handleCreateService(ActivityThread.java:4650)
        at android.app.ActivityThread.access$1700(ActivityThread.java:301)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2185)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:246)
        at android.app.ActivityThread.main(ActivityThread.java:8633)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)

```

**Comment by arjan-s** [comment]
> Another crash after a force close and starting a recording manually in the app.
I didn't think about testing this before, but I see the same behavior.

**Comment by dennisguse** [comment]
@pstorch 
The crash is again a lifecycle problem that I forgot: will do this tomorrow.
The 2nd stacktrace is a result of the previous crash as the TrackRecordingService does not release it's resources properly.
I am working on this, but mocking time is driving me crazy. I will try to continue this next weekend.

**Comment by pstorch** [comment]
The 2nd crash also happens when no track recording is running but you send a Stop via adb. Don't know if it is the same lifecycle problem.

**Comment by dennisguse** [comment]
@pstorch :shrug: but it prevents us from merging this...

**Comment by dennisguse** [comment]
@arjan-s @rgmf @pstorch I need to remove the STOP for now; modifying TrackRecordingActivity to handle this nicely seems to be more complicated than expected.
I will create another issue for this.

**Comment by vanous** [comment]
This is very cool. I tried from Bip button actions in Gadgetbridge and it doesn't work, most likely because we add button_id to the broadcast message. Also, it would be maybe be nice to have `publicapi.ToggleRecording`, but i understand that at this point even the Stop is requiring some extra effort to make it working. Once the Stop starts working, i will take a look on our side :) but do not hurry, i am snowed under some other jobs... :)

**Comment by dennisguse** [comment]
Broadcast? OpenTracks accepts only Intents no Broadcast...
So that shouldn't work.

PS Toggle recording is a good idea!

**Comment by arjan-s** [comment]
@vanous instead of implementing this intent yourself in Gadgetbridge, you can use the interface I'm building [here](https://codeberg.org/Freeyourgadget/Gadgetbridge/pulls/2577)

**Pull request merged**